ports/stm32f4/common-hal/busio/UART.c

```diff
@@ -36,45 +36,455 @@
+        GPIO_InitStruct.Pull = GPIO_PULLUP;
+        GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        GPIO_InitStruct.Alternate = self->tx->altfn_index;
+        HAL_GPIO_Init(pin_port(tx->port), &GPIO_InitStruct);
+    }
+    if (self->rx!=NULL) {
+        GPIO_InitStruct.Pin = pin_mask(rx->number);
+        GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+        GPIO_InitStruct.Pull = GPIO_PULLUP;
+        GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        GPIO_InitStruct.Alternate = self->rx->altfn_index; 
+        HAL_GPIO_Init(pin_port(rx->port), &GPIO_InitStruct);
+    }
+
+    uart_clk_irq_enable(self,USARTx);
+
+    self->handle.Instance = USARTx;
+    self->handle.Init.BaudRate = baudrate;
+    self->handle.Init.WordLength = (bits == 9) ? UART_WORDLENGTH_9B : UART_WORDLENGTH_8B;
+    self->handle.Init.StopBits = (stop>1) ? UART_STOPBITS_2 : UART_STOPBITS_1;
+    self->handle.Init.Parity = (parity==PARITY_ODD) ? UART_PARITY_ODD : 
+                               (parity==PARITY_EVEN) ? UART_PARITY_EVEN :
+                               UART_PARITY_NONE;
+    self->handle.Init.Mode = (self->tx != NULL && self->rx != NULL) ? UART_MODE_TX_RX :
+                             (self->tx != NULL) ? UART_MODE_TX :
+                             UART_MODE_RX;
+    self->handle.Init.HwFlowCtl = UART_HWCONTROL_NONE;
+    self->handle.Init.OverSampling = UART_OVERSAMPLING_16;
+    if (HAL_UART_Init(&self->handle) != HAL_OK)
+    {
+        mp_raise_ValueError(translate("UART Init Error"));
+
+    }
+
+    // Init buffer for rx and claim pins
+    if (self->rx != NULL) {
+        ringbuf_alloc(&self->rbuf, receiver_buffer_size, true);
+        if (!self->rbuf.buf) {
+            mp_raise_ValueError(translate("UART Buffer allocation error"));
+        }
+        claim_pin(rx);
+    }
+    if (self->tx != NULL) {
+        claim_pin(tx);
+    }
+    self->baudrate = baudrate;
+    self->timeout_ms = timeout * 1000;
+
+    //start the interrupt series
+    if ((HAL_UART_GetState(&self->handle) & HAL_UART_STATE_BUSY_RX) == HAL_UART_STATE_BUSY_RX) {
+        mp_raise_ValueError(translate("Could not start interrupt, RX busy"));
+    }
+
+    //start the recieve interrupt chain
+    HAL_NVIC_DisableIRQ(self->irq); //prevent handle lock contention
+    HAL_UART_Receive_IT(&self->handle, &self->rx_char, 1);
+    HAL_NVIC_SetPriority(self->irq, UART_IRQPRI, UART_IRQSUB_PRI);
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    errflag = HAL_OK;
 }
 
 bool common_hal_busio_uart_deinited(busio_uart_obj_t *self) {
-    return 0;
+    return self->tx->pin == mp_const_none;
 }
 
 void common_hal_busio_uart_deinit(busio_uart_obj_t *self) {
+    reset_pin_number(self->tx->pin->port,self->tx->pin->number);
+    reset_pin_number(self->rx->pin->port,self->rx->pin->number);
+    self->tx = mp_const_none;
+    self->rx = mp_const_none;
+    gc_free(self->rbuf.buf);
+    self->rbuf.size = 0;
+    self->rbuf.iput = self->rbuf.iget = 0;
```

Comment:
If this gets called twice, bad things will happen, like gc_free() called twice, etc. So check for deinited before doing all the work.